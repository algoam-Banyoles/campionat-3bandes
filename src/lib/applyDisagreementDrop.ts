import type { SupabaseClient } from '@supabase/supabase-js';

export async function applyDisagreementDrop(
  supabase: SupabaseClient,
  eventId: string,
  playerA: string,
  playerB: string
): Promise<void> {
  const { error } = await supabase.rpc('apply_disagreement_drop', {
    event_id: eventId,
    player_a: playerA,
    player_b: playerB
  });
  if (error) throw new Error(error.message);
}
