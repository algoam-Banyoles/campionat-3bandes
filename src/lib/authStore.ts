import { writable } from 'svelte/store';
import type { Session, User } from '@supabase/supabase-js';
import { supabase } from '$lib/supabaseClient';

export const session = writable<Session | null>(null);
export const user = writable<User | null>(null);
export const loadingAuth = writable(true);

async function init() {
  const { data } = await supabase.auth.getSession();
  session.set(data.session ?? null);
  user.set(data.session?.user ?? null);
  loadingAuth.set(false);

  supabase.auth.onAuthStateChange((_event, sess) => {
    session.set(sess ?? null);
    user.set(sess?.user ?? null);
  });
}
init();
