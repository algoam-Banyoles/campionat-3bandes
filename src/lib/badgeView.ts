import { getPlayerBadges, type VPlayerBadges } from '$lib/playerBadges';

export type BadgeView = {
  label: string;
  text: string;
  className: string;
};

const ACTIVE_BADGE: BadgeView = {
  label: 'Repte actiu',
  text: 'Repte actiu',
  className: 'rounded bg-blue-100 px-1.5 py-0.5 text-xs font-medium text-blue-700'
};

const CHALLENGEABLE_BADGE: BadgeView = {
  label: 'Es pot reptar',
  text: 'Es pot reptar',
  className: 'rounded bg-green-100 px-1.5 py-0.5 text-xs font-medium text-green-700'
};

export function getBadgeView(badge: VPlayerBadges | undefined | null): BadgeView | null {
  if (!badge) return null;
  if (badge.has_active_challenge) {
    return ACTIVE_BADGE;
  }
  if (badge.in_cooldown) {
    return null;
  }
  if (badge.can_be_challenged) {
    return CHALLENGEABLE_BADGE;
  }
  return null;
}

export async function fetchBadgeMap(
  getBadges: () => Promise<VPlayerBadges[]> = getPlayerBadges
): Promise<Map<string, VPlayerBadges>> {
  const list = await getBadges();
  return new Map(list.map((badge) => [badge.player_id, badge]));
}
