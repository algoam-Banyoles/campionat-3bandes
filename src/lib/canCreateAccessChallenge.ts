import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateAccessResult = {
  ok: boolean;
  reason: string | null;
};

export async function canCreateAccessChallenge(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateAccessResult> {
  const { data, error } = await supabase.rpc('can_create_access_challenge', {
    p_event: eventId,
    p_reptador: reptadorId,
    p_reptat: reptatId
  });
  if (error) return { ok: false, reason: error.message };
  const result = data as CanCreateAccessResult[] | null;
  if (!result || result.length === 0) return { ok: false, reason: 'No result' };
  return result[0];
}
