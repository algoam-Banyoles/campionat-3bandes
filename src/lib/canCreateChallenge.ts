import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateChallengeResult = {
  ok: boolean;
  reason: string | null;
  warning: string | null;
};

export async function canCreateChallenge(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateChallengeResult> {
  const { data, error } = await supabase.rpc('can_create_challenge', {
    p_event: eventId,
    p_reptador: reptadorId,
    p_reptat: reptatId
  });
  if (error) return { ok: false, reason: error.message, warning: null };
  const result = data as CanCreateChallengeResult[] | null;
  if (!result || result.length === 0) return { ok: false, reason: 'No result', warning: null };
  return result[0];
}
