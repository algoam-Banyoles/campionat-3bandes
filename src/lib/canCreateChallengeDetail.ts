import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateChallengeDetailResult = {
  ok: boolean;
  reason?: string | null;
};

export async function canCreateChallengeDetail(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateChallengeDetailResult> {
  const { data, error } = await supabase.rpc('can_create_challenge_detail', {
    event_id: eventId,
    reptador_id: reptadorId,
    reptat_id: reptatId
  });
  if (error) {
    return { ok: false, reason: error.message };
  }
  // Supabase RPC may return single object or array
  const result = Array.isArray(data) ? data[0] : data;
  if (!result) {
    return { ok: false, reason: 'No result' };
  }
  return result as CanCreateChallengeDetailResult;
}

