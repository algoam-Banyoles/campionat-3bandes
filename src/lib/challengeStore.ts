// src/lib/challengeStore.ts
import { writable } from 'svelte/store';
import { supabase } from '$lib/supabaseClient';
import { cacheManager } from '$lib/cache/strategies';
// import { performanceMonitor } from '$lib/monitoring/performance';

// Stub temporal mentre TypeScript recompila
const performanceMonitor = {
  measure: async <T>(name: string, category: string, fn: () => Promise<T>, metadata?: any): Promise<T> => {
    return await fn();
  }
};

export type Challenge = {
  id: string;
  event_id: string;
  tipus: 'normal' | 'access';
  reptador_id: string;
  reptat_id: string;
  estat: 'proposat' | 'acceptat' | 'programat' | 'refusat' | 'caducat' | 'jugat' | 'anullat';
  dates_proposades: string[];
  data_proposta: string;
  data_acceptacio?: string | null;
  data_programada?: string | null;
  reprogram_count: number;
  pos_reptador?: number | null;
  pos_reptat?: number | null;
  reptador_nom?: string;
  reptat_nom?: string;
};

export type ChallengeWithPlayers = Challenge & {
  reptador_nom: string;
  reptat_nom: string;
};

export const activeChallenges = writable<ChallengeWithPlayers[]>([]);
export const userChallenges = writable<ChallengeWithPlayers[]>([]);

/**
 * Actualitzar reptes actius amb cache i optimització
 */
export async function refreshActiveChallenges(): Promise<void> {
  try {
    const data = await performanceMonitor.measure(
      'active_challenges_query',
      'database',
      async () => {
        return await cacheManager.get(
          'activeChallenges',
          async () => {
            // Primer obtenir els challenges
            const { data: challenges, error } = await supabase
              .from('challenges')
              .select(`
                id,
                event_id,
                tipus,
                reptador_id,
                reptat_id,
                estat,
                dates_proposades,
                data_proposta,
                data_acceptacio,
                data_programada,
                reprogram_count,
                pos_reptador,
                pos_reptat
              `)
              .in('estat', ['proposat', 'acceptat', 'programat'])
              .order('data_proposta', { ascending: false });

            if (error) {
              console.warn('[challengeStore] refreshActiveChallenges error:', error.message);
              throw new Error(error.message);
            }

            if (!challenges || challenges.length === 0) {
              return [] as ChallengeWithPlayers[];
            }

            // Obtenir noms dels jugadors per separat per evitar duplicació
            const playerIds = Array.from(new Set([
              ...challenges.map(c => c.reptador_id),
              ...challenges.map(c => c.reptat_id)
            ]));

            const { data: players, error: playersError } = await supabase
              .from('players')
              .select('id, nom')
              .in('id', playerIds);

            if (playersError) {
              console.warn('[challengeStore] players query error:', playersError.message);
            }

            // Crear un mapa de jugadors per ID
            const playerMap = new Map<string, string>();
            (players || []).forEach(player => {
              playerMap.set(player.id, player.nom);
            });

            // Processar dades per afegir noms dels jugadors
            const processedData = challenges.map(challenge => ({
              ...challenge,
              reptador_nom: playerMap.get(challenge.reptador_id) || 'Desconegut',
              reptat_nom: playerMap.get(challenge.reptat_id) || 'Desconegut'
            }));

            return processedData as ChallengeWithPlayers[];
          }
        );
      }
    );

    activeChallenges.set(data);

  } catch (error: any) {
    console.warn('[challengeStore] refreshActiveChallenges error:', error.message);
    activeChallenges.set([]);
  }
}

/**
 * Obtenir reptes d'un jugador específic
 */
export async function refreshUserChallenges(playerId: string): Promise<void> {
  try {
    const data = await performanceMonitor.measure(
      'user_challenges_query',
      'database',
      async () => {
        return await cacheManager.get(
          'userChallenges',
          async () => {
            // Primer obtenir els challenges de l'usuari
            const { data: challenges, error } = await supabase
              .from('challenges')
              .select(`
                id,
                event_id,
                tipus,
                reptador_id,
                reptat_id,
                estat,
                dates_proposades,
                data_proposta,
                data_acceptacio,
                data_programada,
                reprogram_count,
                pos_reptador,
                pos_reptat
              `)
              .or(`reptador_id.eq.${playerId},reptat_id.eq.${playerId}`)
              .order('data_proposta', { ascending: false })
              .limit(50); // Limitar per rendiment

            if (error) {
              console.warn('[challengeStore] refreshUserChallenges error:', error.message);
              throw new Error(error.message);
            }

            if (!challenges || challenges.length === 0) {
              return [] as ChallengeWithPlayers[];
            }

            // Obtenir noms dels jugadors per separat
            const playerIds = Array.from(new Set([
              ...challenges.map(c => c.reptador_id),
              ...challenges.map(c => c.reptat_id)
            ]));

            const { data: players, error: playersError } = await supabase
              .from('players')
              .select('id, nom')
              .in('id', playerIds);

            if (playersError) {
              console.warn('[challengeStore] players query error:', playersError.message);
            }

            // Crear un mapa de jugadors per ID
            const playerMap = new Map<string, string>();
            (players || []).forEach(player => {
              playerMap.set(player.id, player.nom);
            });

            // Processar dades per afegir noms dels jugadors
            const processedData = challenges.map(challenge => ({
              ...challenge,
              reptador_nom: playerMap.get(challenge.reptador_id) || 'Desconegut',
              reptat_nom: playerMap.get(challenge.reptat_id) || 'Desconegut'
            }));

            return processedData as ChallengeWithPlayers[];
          },
          `user_${playerId}` // Usar modificador de clau per cache per usuari
        );
      },
      { playerId }
    );

    userChallenges.set(data);

  } catch (error: any) {
    console.warn('[challengeStore] refreshUserChallenges error:', error.message);
    userChallenges.set([]);
  }
}

/**
 * Invalidar cache dels challenges després d'actualitzacions
 */
export function invalidateChallengeCaches(playerId?: string): void {
  cacheManager.invalidate('activeChallenges');
  
  if (playerId) {
    cacheManager.invalidate('userChallenges', `user_${playerId}`);
  }
  
  // També invalidar rankings ja que poden haver canviat
  cacheManager.invalidate('ranking');
}

/**
 * Acceptar un repte i invalidar caches relacionats
 */
export async function acceptChallenge(challengeId: string, playerId?: string): Promise<void> {
  try {
    const { error } = await supabase
      .from('challenges')
      .update({ 
        estat: 'acceptat', 
        data_acceptacio: new Date().toISOString() 
      })
      .eq('id', challengeId);

    if (error) {
      throw new Error(error.message);
    }

    // Invalidar caches i actualitzar dades
    invalidateChallengeCaches(playerId);
    await refreshActiveChallenges();
    
    if (playerId) {
      await refreshUserChallenges(playerId);
    }

  } catch (error: any) {
    console.error('[challengeStore] acceptChallenge error:', error.message);
    throw error;
  }
}

/**
 * Refusar un repte i invalidar caches relacionats
 */
export async function refuseChallenge(challengeId: string, playerId?: string): Promise<void> {
  try {
    const { error } = await supabase
      .from('challenges')
      .update({ estat: 'refusat' })
      .eq('id', challengeId);

    if (error) {
      throw new Error(error.message);
    }

    // Invalidar caches i actualitzar dades
    invalidateChallengeCaches(playerId);
    await refreshActiveChallenges();
    
    if (playerId) {
      await refreshUserChallenges(playerId);
    }

  } catch (error: any) {
    console.error('[challengeStore] refuseChallenge error:', error.message);
    throw error;
  }
}