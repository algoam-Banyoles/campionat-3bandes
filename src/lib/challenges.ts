import type { SupabaseClient } from '@supabase/supabase-js';
import { authFetch } from '$lib/utils/http';

export type ChallengeParticipants = {
  reptador_id: string | null;
  reptat_id: string | null;
};

export function isParticipant(
  mePlayerId: string | null,
  repte: ChallengeParticipants | null | undefined
): boolean {
  if (!mePlayerId || !repte) return false;
  return mePlayerId === repte.reptador_id || mePlayerId === repte.reptat_id;
}

export async function acceptChallenge(supabase: SupabaseClient, id: string): Promise<void> {
  const { error } = await supabase
    .from('challenges')
    .update({ estat: 'acceptat', data_acceptacio: new Date().toISOString() })
    .eq('id', id);
  if (error) throw new Error(error.message);
}

export async function refuseChallenge(supabase: SupabaseClient, id: string): Promise<void> {
  const { error } = await supabase
    .from('challenges')
    .update({ estat: 'refusat' })
    .eq('id', id);
  if (error) throw new Error(error.message);
}

export async function scheduleChallenge(
  _supabase: SupabaseClient,
  id: string,
  isoDate: string
): Promise<void> {
  const res = await authFetch('/campionat-continu/reptes/programar', {
    method: 'POST',
    body: JSON.stringify({ id, data_iso: isoDate })
  });
  const body = await res.json().catch(() => ({}));
  if (!res.ok || body?.ok === false) {
    throw new Error(body?.error || 'Error programant repte');
  }
}


export async function resolveAccessChallenge(
  supabase: SupabaseClient,
  challengeId: string,
  winner: 'reptador' | 'reptat'
): Promise<{ ok: boolean; error?: string }> {
  const { data, error } = await supabase.rpc('resolve_access_challenge', {
    challenge_id: challengeId,
    winner
  });
  if (error) throw new Error(error.message);
  const res = (data as any) ?? { ok: false };
  if (!res.ok) throw new Error(res.error || 'Error resolent repte d\'acc√©s');
  return res;
}

