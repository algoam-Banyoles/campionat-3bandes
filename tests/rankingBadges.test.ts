import { describe, expect, it, vi } from 'vitest';
import { fetchBadgeMap, getBadgeView } from '../src/lib/badgeView';
import type { VPlayerBadges } from '../src/lib/playerBadges';

describe('badge rendering', () => {
  const badgeFixtures: VPlayerBadges[] = [
    {
      event_id: 'event-1',
      player_id: 'p1',
      posicio: 1,
      last_play_date: '2024-01-01',
      days_since_last: 0,
      has_active_challenge: true,
      in_cooldown: false,
      can_be_challenged: false,
      cooldown_days_left: 0
    },
    {
      event_id: 'event-1',
      player_id: 'p2',
      posicio: 2,
      last_play_date: '2024-01-02',
      days_since_last: 1,
      has_active_challenge: false,
      in_cooldown: true,
      can_be_challenged: false,
      cooldown_days_left: 3
    },
    {
      event_id: 'event-1',
      player_id: 'p3',
      posicio: 3,
      last_play_date: '2024-01-03',
      days_since_last: 2,
      has_active_challenge: false,
      in_cooldown: false,
      can_be_challenged: true,
      cooldown_days_left: 0
    }
  ];

  it('maps player badges to the correct view labels', async () => {
    const mockGetBadges = vi.fn<[], Promise<VPlayerBadges[]>>().mockResolvedValue(badgeFixtures);

    const map = await fetchBadgeMap(mockGetBadges);

    expect(mockGetBadges).toHaveBeenCalledTimes(1);

    const activeView = getBadgeView(map.get('p1'));
    const cooldown = map.get('p2');
    const challengeableView = getBadgeView(map.get('p3'));

    expect(activeView).not.toBeNull();
    expect(activeView?.text).toBe('Repte actiu');
    expect(activeView?.label).toBe('Repte actiu');

    expect(cooldown).not.toBeNull();
    expect(cooldown?.in_cooldown).toBe(true);
    expect(cooldown?.cooldown_days_left).toBe(3);
    expect(getBadgeView(cooldown)).toBeNull();

    expect(challengeableView).not.toBeNull();
    expect(challengeableView?.text).toBe('Es pot reptar');
    expect(challengeableView?.label).toBe('Es pot reptar');
  });

  it('produces consistent markup for the three badge states', async () => {
    const map = await fetchBadgeMap(async () => badgeFixtures);
    const html = badgeFixtures
      .map((badge) => {
        if (badge.in_cooldown) {
          const days = badge.cooldown_days_left;
          const article = days === 1 ? 'a' : 'en';
          const unit = days === 1 ? 'dia' : 'dies';
          return `<span class="px-2 py-0.5 text-xs rounded bg-orange-100 text-orange-700">No es pot reptar (falt${article} ${days} ${unit})</span>`;
        }
        const view = getBadgeView(map.get(badge.player_id));
        if (!view) return '';
        return `<span class="${view.className}" aria-label="${view.label}">${view.text}</span>`;
      })
      .join('');

    expect(html).toMatchSnapshot();
  });

  it('returns null when there is no badge information', () => {
    expect(getBadgeView(undefined)).toBeNull();
    expect(getBadgeView(null)).toBeNull();
  });
});
