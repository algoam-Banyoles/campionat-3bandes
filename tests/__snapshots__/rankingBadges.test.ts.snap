// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`badge rendering > produces consistent markup for the three badge states 1`] = `"<span class="rounded bg-blue-100 px-1.5 py-0.5 text-xs font-medium text-blue-700" aria-label="Repte actiu">Repte actiu</span><span class="px-2 py-0.5 text-xs rounded bg-orange-100 text-orange-700">No es pot reptar (falten 3 dies)</span><span class="rounded bg-green-100 px-1.5 py-0.5 text-xs font-medium text-green-700" aria-label="Es pot reptar">Es pot reptar</span>"`;
